tools/saas_promotion_state/saas_promotion_state.py

```diff
@@ -0,0 +1,81 @@
+from __future__ import annotations
+
+from collections.abc import Iterable
+
+from reconcile.openshift_saas_deploy import (
+    QONTRACT_INTEGRATION as OPENSHIFT_SAAS_DEPLOY,
+)
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.typed_queries.saas_files import SaasFile, get_saas_files
+from reconcile.utils.promotion_state import PromotionState
+from reconcile.utils.secret_reader import create_secret_reader
+from reconcile.utils.state import init_state
+
+
+class SaasPromotionState:
+    def __init__(
+        self, promotion_state: PromotionState, saas_files: Iterable[SaasFile]
+    ) -> None:
+        self._promotion_state = promotion_state
+        self._saas_files = saas_files
+
+    def _publisher_ids_for_channel(
+        self, channel: str, saas_files: Iterable[SaasFile]
+    ) -> list[str]:
+        publisher_uids: list[str] = []
+        for saas_file in saas_files:
+            for resource_template in saas_file.resource_templates:
+                for target in resource_template.targets:
+                    if not target.promotion:
+                        continue
+                    for publish_channel in target.promotion.publish or []:
+                        if publish_channel == channel:
+                            publisher_uids.append(
+                                target.uid(
+                                    parent_saas_file_name=saas_file.name,
+                                    parent_resource_template_name=resource_template.name,
+                                )
+                            )
```

Comment:
can there be duplicate values in `publisher_uids`?

Reply:
There shouldn't be duplicates:
- Every saas target has a unique id (theoretical hash collisions possible, but Id neglect this possibility).
- We focus on fetching publishers for a single channel. I.e., the same publisher cannot be in that channel multiple times.